### ISSUE TITLE
Data annotation for date fields

### ISSUE THREAD

**Issue by gusbrs** [plk/biblatex#690]
From https://tex.stackexchange.com/q/410621/105447 .

Biblatex provides the handy feature of data annotation, which allows us to conditionally format fields, items and name parts according to semantic information added at the bibentry level. However, this currently doesn't work for date fields, because `\iffieldannotation` won't work for these fields.

I can think of at least one relevant instance where annotating the date might be relevant, that of attributed dates. It is common in references to indicate an attributed field by putting it into brackets. If a book comes with no date printed, but we happen to know the date (so it is not "uncertain", nor "circa"), we can indicate that by putting the date within brackets. Data annotation could be used to do that, it would indeed be the best way to do so (afaik, the only alternative would be to add the brackets in the `year` field).

So I kindly request the extension of the data annotation feature to date fields.

The MWE below ilustrates the problem:

```
\documentclass{article}

\usepackage[style=authoryear]{biblatex}
\usepackage{filecontents}

\begin{filecontents}{\jobname.bib}
@article{gillies,
  author       = {Gillies, Alexander},
  title        = {Herder and the Preparation of {Goethe's} Idea of World Literature},
  journaltitle = {Publications of the English Goethe Society},
  date         = {1933},
  date+an      = {=attributed},
  series       = {newseries},
  volume       = {9},
  volume+an    = {=attributed},
  pages        = {46-67},
}
\end{filecontents}

\addbibresource{\jobname.bib}

% This works, but affects all dates
\DeclareFieldFormat{date}{\mkbibbrackets{#1}}
\DeclareFieldFormat{labeldate}{\mkbibbrackets{#1}}

% This doesn't work, for the conditional is not recognized
%\DeclareFieldFormat{date}{\iffieldannotation{attributed}{\mkbibbrackets{#1}}{#1}}
%\DeclareFieldFormat{labeldate}{\iffieldannotation{attributed}{\mkbibbrackets{#1}}{#1}}

% It works with the conditional for volume, for example
\DeclareFieldFormat[article]{volume}{\iffieldannotation{attributed}{\mkbibbrackets{#1}}{#1}}

\begin{document}

\autocite{gillies}

\printbibliography

\end{document}
```

**Comment by moewew** [comment]
Have a look at https://github.com/plk/biblatex/compare/dev...moewew:moreannotation. The branch defines `\ifdateannotation` exactly as in the TeX.SX answer.

**Comment by gusbrs** [comment]
@moewew , I've played with it here and it looks great. Thank you very much once again. Appreciated. Putting it all together, I guess I'll be anxiously waiting for the next release.  :)

**Comment by moewew** [comment]
I'd also like to implement a version of `\iffieldannotation`, `\ifitemannotation` and `\ifpartannotation` that don't need to be used in `\currentfield` scope where the relevant field and item number is provided as argument. But I can't think of a good name. Ideas?

**Comment by plk** [comment]
I wonder in such cases if we could detect i in a current field scope and if so gobble the args and get them from the scope. That way we don’t need new names? That is, only require the arguments if not in current field scope?

**Comment by moewew** [comment]
In that case I would have used a star for autodetect and no star for an argument, but then we would have to change the syntax of existing commands, and that might not be a great idea.

**Comment by gusbrs** [comment]
If all hope to put everything under the same hood is lost, you might consider `ifthefieldannotation`, `iftheitemannotation` etc. "the" opposed to "current". It's not great, but it is the best I could come up with. Anyway, if it is possible, I much prefer the approach you are both attempting, of using one name/command which behaves according to context.

**Comment by moewew** [comment]
`biblatex-oxref` already uses `\iffieldannotation` and `\ifitemannotation` so we can't alter the syntax in a backwards incompatible way.

**Comment by moewew** [comment]
The following should be possible

- `\iffieldannotation*{field}{annotation}` is given the field explicitly
- `\iffieldannnotation{annotation}` auto-detects the field

The downside to this approach is that it reverses my intuition w.r.t. the starred version: I would have expected the starred version to be more automatic (compare commands defined with `\DeclarePairedDelimiter`). And the starred and unstarred version have a different number of arguments, I'm not sure if that is best practice.

**Comment by moewew** [comment]
Another alternative would be an optional argument

- `\iffieldannotation[field]{annotation}` is given the field explicitly
- `\iffieldannnotation{annotation}` auto-detects the field

At the moment I like this approach most, but it is different from what other `\iffield...`macros do. But since `\iffieldannotation` already deviates from the norm by not having different versions for `name` and `list` anyway, we might be fine.

**Comment by moewew** [comment]
The approach using optional arguments looked most promising to me so there is https://github.com/plk/biblatex/compare/dev...moewew:moreannotation now.

**Comment by plk** [comment]
Yes, looks nice - feel free to merge

**Comment by moewew** [comment]
Will wait a bit before I merge. I have the habit of discovering I made some sort of stupid mistake just minutes after I merged or committed a bit of code.

**Comment by moewew** [comment]
#692 is merged now. Testing would of course be appreciated if anyone has good test cases.

**Comment by gusbrs** [comment]
@moewew , I tried things out in a set of plausible situations (bib drivers, cite commands) and, as far as I cant tell, it looks great. Thanks a lot!

**Comment by moewew** [comment]
Thank you very much for checking. If you find any bugs, feel free to report them here.

**Comment by gusbrs** [comment]
@moewew , I think I found a corner case for the date annotation which does not behave as expected.
If I use the annotation within an `\entrydata` group, the date printed within this group will receive the annotation of the "surrounding environment". I'm not sure if this belongs here, in the "save date fields" issue, or in a new one. Please advise.

Consider the following MWE:

```
\documentclass{article}

\usepackage[%
    english,
%    ngerman,
%    french,
%    brazilian,
    ]{babel}

\usepackage[style=authoryear]{biblatex}

\usepackage{filecontents}

\begin{filecontents}{\jobname.bib}
@book{mybook,
  author    = {Author},
  title     = {Book Title},
  date      = {2018-11-07},
}
@customa{myentry,
  author    = {Buthor},
  title     = {My entry's title},
  subtitle  = {Below comes the testing},
  date      = {2017-04-05},
  date+an   = {=attributed},
  xref      = {mybook},
}
\end{filecontents}

\addbibresource{\jobname.bib}

\DeclareBibliographyDriver{customa}{%
  \printnames{author}%
  \setunit{\printdelim{nametitledelim}}%
  \printfield{title}%
  \setunit{\subtitlepunct}%
  \printfield{subtitle}%
  \newunit\newblock
  \printdate
  \newunit\newblock
  \entrydata*{\thefield{xref}}{%
    \usebibmacro{cite}%
  }
}

\DeclareFieldFormat{date}{\ifdateannotation{date}{attributed}{\mkbibbrackets{#1}}{#1}}
\DeclareFieldAlias{labeldate}{date}

\begin{document}
\fullcite{mybook}

\fullcite{myentry}

\printbibliography
\end{document}
```

With current dev version, it results in:

![mwe](https://user-images.githubusercontent.com/19410606/36618344-97f2bda0-18c9-11e8-87ba-b0a27d79e610.png)

**Comment by moewew** [comment]
Thank you for reporting this. 

It gets even worse. With
```latex
\DeclareBibliographyDriver{customa}{%
  \printnames{author}%
  \setunit{\printdelim{nametitledelim}}%
  \printfield{title}%
  \newunit\newblock
  \printdate
  \newunit\newblock
  \entrydata*{\thefield{xref}}{%
    \printdate
    \newunit\newblock
    \printsaveddate
  }
}
```
We get
> Buthor. *My entry’s title.* [Apr. 5, 2017]. [Nov. 7, 2018]. Apr. 5, 2017

I think this is because annotations are not really part of the data model considered by `biblatex`. This could be tricky.

**Comment by moewew** [comment]
Please have a look at https://github.com/plk/biblatex/compare/dev...moewew:saveannotations

```latex
\documentclass{article}

\usepackage[style=authoryear]{biblatex}

\usepackage{filecontents}

\begin{filecontents}{\jobname.bib}
@book{mybook,
  author    = {Author},
  title     = {Book Title},
  date      = {2018-11-07},
}
@customa{myentry,
  author    = {Buthor},
  title     = {My entry's title},
  date      = {2017-04-05},
  date+an   = {=attributed},
  xref      = {mybook},
}
\end{filecontents}

\addbibresource{\jobname.bib}

\makeatletter
\DeclareBibliographyDriver{customa}{%
  \printnames{author}%
  \setunit{\printdelim{nametitledelim}}%
  \printfield{title}%
  \newunit\newblock
  \printdate
  \newunit\newblock
  \entrydata*{\thefield{xref}}{%
    \printdate
    \newunit\newblock
    \printsaveddate
  }
}

\DeclareFieldFormat{date}{\ifdateannotation{date}{attributed}{\mkbibbrackets{#1}}{#1}}
\DeclareFieldFormat{saveddate}{\ifdateannotation{saveddate}{attributed}{\mkbibbrackets{#1}}{#1}}
\DeclareFieldAlias{labeldate}{date}


\begin{document}
\nocite{*}
\printbibliography
\end{document}
```
now gives
![kjhjjkhhkhjkh](https://user-images.githubusercontent.com/6755835/36628828-33617b9e-194e-11e8-9790-cbbfb4cd0a11.png)

**Comment by gusbrs** [comment]
@moewew , thank you for the prompt response.

I found one (or two) issue(s) with this commit. The first is that the saved annotations don't seem to work as expected, in case you nest `\entrydata`s.

```
\documentclass{article}

\usepackage[style=authoryear]{biblatex}

\usepackage{filecontents}

\begin{filecontents}{\jobname.bib}
@book{mybook,
  author    = {Author},
  title     = {Book Title},
  date      = {2018-11-07},
}
@customa{myentry,
  author    = {Buthor},
  title     = {My entry's title},
  date      = {2017-04-05},
  date+an   = {=attributed},
  xref      = {mybook},
}
\end{filecontents}

\addbibresource{\jobname.bib}

\makeatletter
\DeclareBibliographyDriver{customa}{%
  \printnames{author}%
  \setunit{\printdelim{nametitledelim}}%
  \printfield{title}%
  \newunit\newblock
  \printdate
  \newunit\newblock
  \entrydata*{\thefield{xref}}{%
    \printdate
    \newunit\newblock
    \printsaveddate%
    \entrydata*{\thefield{savedentrykey}}{%
      \printdate
      \newblock
      \printsaveddate
    }
  }
}

\DeclareFieldFormat{date}{\ifdateannotation{date}{attributed}{\mkbibbrackets{#1}}{#1}}
\DeclareFieldFormat{saveddate}{\ifdateannotation{saveddate}{attributed}{\mkbibbrackets{#1}}{#1}}
\DeclareFieldAlias{labeldate}{date}


\begin{document}
\nocite{*}
\printbibliography
\end{document}
```

Results in:

![mwe](https://user-images.githubusercontent.com/19410606/36631535-c162ff30-1958-11e8-9c56-c91cd32a99d8.png)

The (possible) second one happens in my test document (in my real case settings). I have a similar structure in some of my drivers as the one in the above mwe, and I use within the `entrydata` the `cite` macro. Now, only in entries of mine which have a date annotation such as this one, a spurious space before the call to `cite` is introduced. A spurious space I could not trace and also could not reproduce in a smaller example. So the fault might well be with me. But I can tell you the space only emerges on entries with date annotations, and it is also not there with the current dev version.

**Comment by moewew** [comment]
Thank you for testing.

A fix for this is now in https://github.com/plk/biblatex/compare/dev...moewew:saveannotations

There was an uncommented line end, maybe that was the cause of the unwanted space. If it's still there even with the commit adding the `%`, I would need to see an MWE.

**Comment by gusbrs** [comment]
@moewew , all tests look very fine here now. And the space is indeed gone. Thank you very much.

**Comment by gusbrs** [comment]
Btw, I think this is a proper opportunity for a more general thank you.

The fact that I found this issue in my current working document means that I am now a happy user of a style agnostic (or almost so) citation structure for primary / archival sources to work with biblatex's standard styles. I'm a economic historian by trade and I rely heavily on this kind of sources. And I'm unaware of any bibliography management platform, other than biblatex, which would allow me to get even close to such a structure. Biblatex not only enabled me to do so, but did so deftly and robustly. For someone with my outlook, LaTeX has many qualities, but biblatex stands among them very highly. So, through @plk, I'd like to extend my warmest thank you to all involved. From where I stand, the work you do here is really, really appreciated.

@moewew in particular, actively and generously assisted me when I stumbled in this attempt of mine, which would doubtfully be standing as it is without this assistance.

Thank you very much!

**Comment by moewew** [comment]
The changes have been merged to dev. Since the Biber release and with it also the release of `biblatex` is still blocked by missing binaries for certain platforms, this may make it into `biblatex` 3.11 after all.

General testing of the current dev branch would be appreciated.

**Comment by gusbrs** [comment]
@moewew , this looks good on dev too. Thank you.

In general, I took a look (admittedly not as thoroughly as originally) to the issues pertaining to this release in which I was somehow involved (this one, #684 and #583).
As far as I can tell, it all behaves as expected.

I noticed one change to the date annotation from my previous test, but it looks for the best. When I originally tested, annotated circa dates would be rendered as, e.g. `ca. [1985]`. Somewhere along the way this changed and now they are rendered `[ca. 1985]`. One could debate which is nicer for this particular use case. But from a general perspective within biblatex, the circa part should obviously be included in the annotation, as it is part of the date. There may come a day when date parts annotations are supported, as they are for names. But that's another matter altogether. As it currently stands, it seems better than initially.

I also noticed another new string, which I will report on #583.

Altogether, it all looks shinning to me.
